### ISSUE TITLE
Addon-controls: Feature request to allow user-defined controls

### ISSUE THREAD

**Issue by stefanhamburger** [storybookjs/storybook#11486]
We are currently documenting our React components in Storybook, and decided to go straight to version 6.0 to make use of the new Args and Controls addons. During that work, we realised that the [choice of controls](https://github.com/storybookjs/storybook/blob/next/addons/controls/README.md#control-annotations) is very limited.

Here are some examples:
- We want users to select from a pre-defined list of colors. The `"select"` control can only display text but we want to preview the color next to it.
- We want users to select from a list of icons. The control should display a list of all icons and clicking on an icon selects it.
- One prop may allow for `string[]`, so we want to allow the user to add strings to the list, or remove existing strings.

Any chance to allow users to specify custom controls in `argTypes` that are not built into `addon-controls`?

Here's how this could look like in React code, where `ColorPicker` is a custom control we'd build:

```tsx
const ColorPicker = ({ value, setValue }) => {
  return (
    <select value={value} onChange={setValue}>
      <option>red</option>
      <option>blue</option>
    </select>
  );
}

export default {
  title: "ExampleComponent",
  component: ExampleComponent,
  argTypes: {
    color: {
      control: ColorPicker,
    },
  },
};
```

**Comment by ryaninvents** [comment]
Seconded. Would love to add a gradient editor and a rich text editor without requiring changes to the core controls package.

**Comment by ryaninvents** [comment]
I'd be happy to start working on this. I'd like to add a gradient editor and rich text editor to my Storybook's controls without having to modify the core controls package.

**Comment by Rycochet** [comment]
One generic thing that would be good is nested controls. For arrays and tuples it is relatively simple (a tuple is a fixed length array) - in my use case I'd like to be able to have a `[number, number]` that can get the normal range options on it. Object types would likely be a bit more complicated - though the current way of pasting some JSON is pretty good for most situations.

I've not looked at the source, but I feel that making the recognition and control both generic, so certain ones can try to identify then including user-defined controls that support it (the aforementioned array type might be useful etc)

**Comment by hydrosquall** [comment]
Copying @ndelangen 's suggestion from another thread to make these user-defined controls lazy-loaded: 

https://github.com/storybookjs/storybook/pull/12685#issuecomment-711761883

**Comment by ndelangen** [comment]
@stefanhamburger @hydrosquall @Rycochet @ryaninvents super excited to see ya'll helping 🙇 

@shilman I'm concerned about:

```js
export default {
  title: "ExampleComponent",
  component: ExampleComponent,
  argTypes: {
    color: {
      control: ColorPicker,
    },
  },
};
```

If the actual control component is specified in the story (preview), but the control component itself is displayed in an addon (manager) aren't we opening ourselves up to a world of hurt?

Should we not do this instead:

```js
// story
export default {
  title: "ExampleComponent",
  component: ExampleComponent,
  argTypes: {
    color: {
      control: 'ControlColorPicker',
    },
  },
};
```

```js
// some addon
addons.register('AddonControlsColorPicker', api => {
  addons.add('AddonControlsColorPicker', {
    type: addons.types.CONTROL,
    render: ColorPicker,
  }
})
```

**Comment by Rycochet** [comment]
@ndelangen I'm with you there - it's not so much a security thing, but being able to keep the definition and usage separate makes it easier to debug things. I do feel that it might be a little bit much for simple uses - but given a suitably "powerful" api for registering and using them it could probably reduce the need for actually doing things like that. Definitely needs a nice and simple example however it gets added so that people don't get confused over what to do, and what it can achieve ;-)

**Comment by shilman** [comment]
@ndelangen Yep, the user/addon needs to register a mapping from string to React component in both the preview and the manager, since the control will be rendered in both the addons panel AND in the docs page. The user would then reference that control by string in the argTypes annotation.

**Comment by shilman** [comment]
Strawman spec:
- [ ] Document existing control API
- [ ] Add control registration to addons API (both manager & preview)
- [ ] Convert built-in controls to use registration mechanism (then they can be overridden!)
- [ ] Create a sample custom control in `official-storybook`

**Comment by ndelangen** [comment]
I'd be happy to assist in the creation of this

**Comment by savgrace** [comment]
I'd love to see OTB support for this as well.

**Comment by stubar** [comment]
For objects rather than editing a json string what about some kind of recursive docgen driven form.  The UX would be hard to fathom admittedly.  My immediate thought is you could click an edit CTA (where the json string is today) and the whole form is replaced by the child object's fields, breadcrumbs could allow for navigating back to parent.

**Comment by CaptainN** [comment]
There are sophisticated form generation tools, like [uniforms](https://uniforms.tools/). For most of my needs, I'd like to be able to simply pass a configured Schema in for particular properties with deep data structures, and have that form open in a dialog or something similar.

**Comment by nathanielks** [comment]
> One generic thing that would be good is nested controls.

I would also be interested in adding controls for args with nested properties.

**Comment by ndelangen** [comment]
@nathanielks yes, that has come up a few time in discussions!

In particular in combination with args being composed from other stories!

**Comment by davidpfahler** [comment]
It would be really great to have custom control types. For example, I'd like to create a color control type for tailwindcss classes, specifically for color classes, so you can use a color-picker-style control type, but use tailwindcss classes.

What is holding this up? How can I help?

**Comment by lvndry** [comment]
It sounds like a very nice feature to have, I'd love to help make this happen :)

**Comment by harleyharl** [comment]
I'm keen for this too! Any new update?

**Comment by markrocks** [comment]
I would love this as well1 +1

**Comment by nataliepina** [comment]
Running into issues with this currently. It would be a great feature to have!

**Comment by gereltod-g** [comment]
I would love to use it.

**Comment by adamlohner** [comment]
@shilman @ndelangen can y'all confirm if this is anywhere on the roadmap?

**Comment by franktopel** [comment]
It's been almost two years, and this is a very common need... How long is it gonna take?

**Comment by Klowes** [comment]
I'd appreciate this feature, I think you should just open this up for us.

just let us straight register components to be used as controls, I'm writing a Vue app, but I'd gladly do some react like it's 2018 if that was supported 😛

**Comment by pseudotsuga-fir** [comment]
I'm surprised it's not yet something you can do in storybook! Definitely worth bumping priority up on.

**Comment by i-am-the-slime** [comment]
What is it with the entitledness in the comments? If you want this functionality, create a PR that adds it or [create your own](https://storybook.js.org/docs/react/addons/addons-api) custom controls addon.

**Comment by Klowes** [comment]
@i-am-the-slime idk about you but my company is def not paying me to build out storybook

**Comment by GideonMax** [comment]
@i-am-the-slime with all due respect, some people don't want to have to learn how addons work, then understand how the controls addon works internally, then change it to allow for custom controls, then go through the process of a PR

**Comment by zb-sj** [comment]
I've gathered a [dirty workaround](https://gist.github.com/zb-sj/3a9731ced7e697632ac2c6d774936599) while #21566 is on the way.

This just injects invokable button to the controls.

I might be able to generalize the implementation to add any custom controls and publish proper addon package if there is enough demand.

**Comment by zb-sj** [comment]
> Very nice, thanks. Is there also a solution for Angular? I was thinking about using Angular Material with mat-select and mat-autocomplete (so you can for example add icons or prefix to select input)

I'm afraid the Storybook UI itself is built with React.
It would be difficult and not feasible to use Angular component on React

**Comment by gotwig** [comment]
> > Very nice, thanks. Is there also a solution for Angular? I was thinking about using Angular Material with mat-select and mat-autocomplete (so you can for example add icons or prefix to select input)
> 
> I'm afraid the Storybook UI itself is built with React. It would be difficult and not feasible to use Angular component on React

Alright. What im looking for is a select with html options, so we can for example show icons in front of option names :)

**Comment by dmamontov** [comment]
@zb-sj Hello, it seems to me that [gist](https://gist.github.com/zb-sj/3a9731ced7e697632ac2c6d774936599) is missing constants and InvokerControl. Can I please see them somehow? otherwise I don’t see the full picture. I would like to implement custom controls.

**Comment by samvv** [comment]
If it's feasible I'd like to look into it. I really need this and I have some spare time. I'll try to continue where @zb-sj stopped.

> There are sophisticated form generation tools, like [uniforms](https://uniforms.tools/). For most of my needs, I'd like to be able to simply pass a configured Schema in for particular properties with deep data structures, and have that form open in a dialog or something similar.

That's not really related to this issue, I believe. This issue isn't about adding more built-in controls (like the ones you are describing), but about being able to write your own for whatever purpose you want.

**Comment by TheBit** [comment]
Hi, any news here?

**Comment by samvv** [comment]
Hi,

Sorry, no. I've since moved on to other projects and languages. Forgot what was holding this back. Iirc it had something to do with the testing infrastructure.

Sam

-------- Original Message --------

**Comment by Buslowicz** [comment]
Is this thing happening any time soon? I am working on a storybook for my project, where I would like to use translation keys for labels in stories, and for that it would be great to get sort of autocomplete input, so that it lets picking existing key or entering a free text. This feature would be perfect for this, though would love that to be not React-only.

**Comment by craigrileyuk** [comment]
Yeah. This would be pretty useful.

Doesn't even matter if it's React only, being able to extend the available control types would be pretty cool.

Like for example, we have a lot of args that are class strings. We'd like to be able to turn this into a text entry box with previous classes rendered as 'chips' with a delete button. 

Could likely code this if there was an API surface for adding control types.

**Comment by paaulorobson** [comment]
I would like to implement a custom select group control, the current storybook select does not allow me to customize it

**Comment by mchestnut** [comment]
Our company just implementing Storybook and this is something that has come up immediately. We would love to create custom controls for the specific use case of displaying color swatches in a limited list of options in a select.
